File: features/fixtures/mazerunner/cxx-scenarios/src/main/java/com/bugsnag/android/mazerunner/scenarios/CXXSessionInfoCrashScenario.java
package com.bugsnag.android.mazerunner.scenarios;

import com.bugsnag.android.Bugsnag;
import com.bugsnag.android.Configuration;
import com.bugsnag.android.TestHarnessHooksKt;

import android.content.Context;
import android.os.Handler;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class CXXSessionInfoCrashScenario extends Scenario {

    static {
        System.loadLibrary("cxx-scenarios");
    }

    public native int crash(int value);

    private Handler handler = new Handler();

    public CXXSessionInfoCrashScenario(@NonNull Configuration config,
                                       @NonNull Context context,
                                       @Nullable String eventMetadata) {
        super(config, context, eventMetadata);
        config.setAutoTrackSessions(false);
    }

    @Override
    public void startScenario() {
        super.startScenario();
        String metadata = getEventMetadata();

        Bugsnag.startSession();
        TestHarnessHooksKt.flushAllSessions();
        Bugsnag.notify(new Exception("For the first"));
        Bugsnag.notify(new Exception("For the second"));
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                crash(3837);
            }
        }, 2500);
    }
}

File: features/fixtures/mazerunner/cxx-scenarios/src/main/java/com/bugsnag/android/mazerunner/scenarios/CXXSessionInfoCrashScenario.kt
package com.bugsnag.android.mazerunner.scenarios

import android.content.Context
import com.bugsnag.android.Bugsnag
import com.bugsnag.android.Configuration
import com.bugsnag.android.createDefaultDelivery
import com.bugsnag.android.mazerunner.InterceptingDelivery
import java.util.concurrent.atomic.AtomicInteger

class CXXSessionInfoCrashScenario(
    config: Configuration,
    context: Context,
    eventMetadata: String?
) : Scenario(config, context, eventMetadata) {

    private val deliveryCount = AtomicInteger(0)

    init {
        System.loadLibrary("cxx-scenarios")

        config.delivery = InterceptingDelivery(createDefaultDelivery()) {
            when (deliveryCount.incrementAndGet()) {
                1 -> Bugsnag.notify(Exception("For the first"))
                2 -> Bugsnag.notify(Exception("For the second"))
                3 -> crash(3837)
            }
        }
    }

    external fun crash(value: Int): Int

    override fun startScenario() {
        super.startScenario()
        Bugsnag.startSession()
    }
}
